File: Coding_Exercise/java_exercise/longestValidParentheses.java
public class longestValidParentheses {
    public int longestValidParentheses(String s) {
        // Start typing your Java solution below
        // DO NOT write main() function
        if (s.length() == 0)
            return 0;
            
        int left =0;
        int ret = 0;
        Stack<Integer> si = new Stack<Integer>();            
        for (int i =0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                si.push(i);
            } else {
                if (!si.isEmpty()) {
                    si.pop();  //BUG
                    int r = 0;
                    if (si.isEmpty()) {
                        r = i-left+1;                       
                    } else {
                        r = i-si.peek();
                    }
                    ret = Math.max(ret, r);
                } else {
                    left = i+1;
                    
                }
            }
        }
        return ret;
    }
}

File: Coding_Exercise/LeetCode/LongestValidParentheses.cpp
	int longestValidParentheses(string s) {
		// Start typing your C/C++ solution below
		// DO NOT write int main() function

		int ret = 0;
		if (s.empty())
			return ret;

		stack<int> cstack;
		int length_sofar = 0;
		int left = -1;
		for (int i =0; i < s.size(); ++i) {
			if (s[i] == '(')
				cstack.push(i) ;
			else {
				if (!cstack.empty()) {
					cstack.pop();
					if (cstack.empty())
						length_sofar = i- left;
					else 
						length_sofar = i-cstack.top();
					ret = max(ret, length_sofar);
				} else {
					left = i; //reset when stack is empty(important)
				}      

			}
		}
		return ret;

	}
